contracts/external/cw-payroll-factory/src/contract.rs

```diff
@@ -0,0 +1,325 @@
+
+pub(crate) const CONTRACT_NAME: &str = "crates.io:cw-payroll-factory";
+pub(crate) const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+pub const INSTANTIATE_CONTRACT_REPLY_ID: u64 = 0;
+pub const DEFAULT_LIMIT: u32 = 10;
+pub const MAX_LIMIT: u32 = 50;
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    cw_ownable::initialize_owner(deps.storage, deps.api, msg.owner.as_deref())?;
+
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+    Ok(Response::new()
+        .add_attribute("method", "instantiate")
+        .add_attribute("creator", info.sender))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::Receive(msg) => execute_receive_cw20(env, deps, info, msg),
+        ExecuteMsg::InstantiateNativePayrollContract {
+            instantiate_msg,
+            code_id,
+            label,
+        } => {
+            execute_instantiate_native_payroll_contract(deps, info, instantiate_msg, code_id, label)
+        }
+        ExecuteMsg::UpdateOwnership(action) => execute_update_owner(deps, info, env, action),
+    }
+}
+
+pub fn execute_receive_cw20(
+    _env: Env,
+    deps: DepsMut,
+    info: MessageInfo,
+    receive_msg: Cw20ReceiveMsg,
+) -> Result<Response, ContractError> {
+    // Only accepts cw20 tokens
+    nonpayable(&info)?;
+
+    let msg: ReceiveMsg = from_binary(&receive_msg.msg)?;
+
+    // Save instantiator info for use in reply (cw20 sender in this case)
+    let sender = deps.api.addr_validate(&receive_msg.sender)?;
+    TMP_INSTANTIATOR_INFO.save(deps.storage, &sender)?;
+
+    match msg {
+        ReceiveMsg::InstantiatePayrollContract {
+            instantiate_msg,
+            code_id,
+            label,
+        } => instantiate_contract(deps, info, instantiate_msg, code_id, label),
+    }
+}
+
+pub fn execute_instantiate_native_payroll_contract(
+    deps: DepsMut,
+    info: MessageInfo,
+    instantiate_msg: PayrollInstantiateMsg,
+    code_id: u64,
+    label: String,
+) -> Result<Response, ContractError> {
+    // Save instantiator info for use in reply
+    TMP_INSTANTIATOR_INFO.save(deps.storage, &info.sender)?;
+
+    instantiate_contract(deps, info, instantiate_msg, code_id, label)
+}
+
+pub fn instantiate_contract(
+    deps: DepsMut,
+    info: MessageInfo,
+    instantiate_msg: PayrollInstantiateMsg,
+    code_id: u64,
+    label: String,
+) -> Result<Response, ContractError> {
+    // Check sender is contract owner if set
+    let ownership = cw_ownable::get_ownership(deps.storage)?;
+    if let Some(owner) = ownership.owner {
+        if info.sender != owner {
+            return Err(ContractError::Unauthorized {});
+        }
```

Comment:
currently if an owner is specified, no cw20 vesting contracts can be instantiated as `info.sender` is the cw20 address here.